File: examples/canonical-auth/java/src/test/java/dev/tenantcore/auth/MeControllerTest.java
package dev.tenantcore.auth;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.web.servlet.MockMvc;

import dev.tenantcore.auth.auth.JwtService;
import dev.tenantcore.auth.domain.InMemoryStore;
import dev.tenantcore.auth.domain.User;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
import static org.hamcrest.Matchers.*;

@SpringBootTest
@AutoConfigureMockMvc
public class MeControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private JwtService jwtService;

    @Autowired
    private InMemoryStore store;

    @Test
    public void testMeResponse_ContainsRequiredFields() throws Exception {
        User user = store.getUserByEmail("admin@acme.example");
        String token = jwtService.issueToken(user, store);

        mockMvc.perform(get("/me").header("Authorization", "Bearer " + token))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.sub").value(user.id()))
            .andExpect(jsonPath("$.email").value(user.email()))
            .andExpect(jsonPath("$.tenant_id").value(user.tenantId()))
            .andExpect(jsonPath("$.tenant_role").value("tenant_admin"))
            .andExpect(jsonPath("$.permissions").isArray())
            .andExpect(jsonPath("$.groups").isArray())
            .andExpect(jsonPath("$.allowed_routes").isArray())
            .andExpect(jsonPath("$.issued_at").exists())
            .andExpect(jsonPath("$.expires_at").exists());
    }

    @Test
    public void testMeResponse_AllowedRoutesFiltered() throws Exception {
        User user = store.getUserByEmail("alice@acme.example");
        String token = jwtService.issueToken(user, store);

        mockMvc.perform(get("/me").header("Authorization", "Bearer " + token))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.allowed_routes[?(@.path == '/users' && @.method == 'GET')]").exists())
            // Should not have admin route
            .andExpect(jsonPath("$.allowed_routes[?(@.path == '/admin/tenants')]").doesNotExist());
    }

    @Test
    public void testMeResponse_SuperAdminShape() throws Exception {
        User user = store.getUserByEmail("superadmin@tenantcore.dev");
        String token = jwtService.issueToken(user, store);

        mockMvc.perform(get("/me").header("Authorization", "Bearer " + token))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.tenant_id").value(nullValue()))
            .andExpect(jsonPath("$.tenant_role").value("super_admin"))
            .andExpect(jsonPath("$.groups").isEmpty())
            .andExpect(jsonPath("$.allowed_routes[?(@.path == '/admin/tenants')]").exists())
            .andExpect(jsonPath("$.allowed_routes[?(@.path == '/users')]").doesNotExist());
    }
}

File: examples/canonical-auth/elixir/test/tenantcore_auth/me_controller_test.exs
defmodule TenantcoreAuth.MeControllerTest do
  use TenantcoreAuth.ConnCase, async: true

  alias TenantcoreAuth.Auth.TokenService
  alias TenantcoreAuth.Domain.InMemoryStore

  setup do
    System.put_env("TENANTCORE_TEST_SECRET", "tenantcore-test-secret")
    :ok
  end

  defp auth_header(email) do
    user = InMemoryStore.get_user_by_email(email)
    [{"authorization", "Bearer " <> TokenService.issue_token(user, InMemoryStore)}]
  end

  test "/me returns the required fields", %{conn: conn} do
    conn = Enum.reduce(auth_header("admin@acme.example"), conn, fn {key, value}, acc -> put_req_header(acc, key, value) end)
    conn = get(conn, "/me")
    body = json_response(conn, 200)

    for key <- ~w[sub email display_name tenant_id tenant_name tenant_role groups permissions acl_ver allowed_routes guide_sections issued_at expires_at] do
      assert Map.has_key?(body, key)
    end
  end

  test "/me filters allowed_routes", %{conn: conn} do
    conn = Enum.reduce(auth_header("bob@acme.example"), conn, fn {key, value}, acc -> put_req_header(acc, key, value) end)
    body = conn |> get("/me") |> json_response(200)
    paths = Enum.map(body["allowed_routes"], & &1["path"])
    assert "/users" in paths
    refute "/admin/tenants" in paths
  end

  test "super_admin /me shape", %{conn: conn} do
    conn = Enum.reduce(auth_header("superadmin@tenantcore.dev"), conn, fn {key, value}, acc -> put_req_header(acc, key, value) end)
    body = conn |> get("/me") |> json_response(200)
    assert body["tenant_role"] == "super_admin"
    assert body["tenant_id"] == nil
    assert body["groups"] == []
  end
end
